File: SourceCode/AspCoreVersion/src/ShopAware.Core/Utilities/Utilities.cs
using System;
using System.ComponentModel;
using System.Globalization;
using System.Linq;
using System.Reflection;
using Newtonsoft.Json.Linq;
using ShopAware.Core.Attributes;
using ShopAware.Core.DataObjects;

namespace ShopAware.Core
{
    public sealed class Utilities
    {
        #region Public Methods

        /// <summary>
        ///     Fixes String values
        /// </summary>
        /// <param name="str"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public static string AddBackslash(string str)
        {
            if (str != null && !str.EndsWith("\\"))
            {
                str += "\\";
            }

            return str;
        }

        /// <summary>
        ///     Fixes String values
        /// </summary>
        /// <param name="str"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public static string AddForwardSlash(string str)
        {
            if (str != null && !str.EndsWith("/"))
            {
                str += "/";
            }

            return str;
        }

        /// <summary>
        ///     Converts a string date to a Date Object
        /// </summary>
        /// <param name="dateString">21050102</param>
        /// <param name="dateFormat">yyyyMMdd</param>
        /// <returns></returns>
        /// <remarks></remarks>
        public static DateTime ConvertDateStringToDate(string dateString, string dateFormat)
        {
            if (string.IsNullOrEmpty(dateString))
            {
                return default(DateTime);
            }

            var result = default(DateTime);

            switch (dateFormat)
            {
                case "yyyyMMdd":
                    result = DateTime.ParseExact(dateString, "yyyyMMdd", CultureInfo.CurrentCulture);
                    break;
            }

            return result;
        }

        public static string GetEnumAttribute<T>(object value)
        {
            var fieldInfo = value.GetType().
                                  GetField(value.ToString());
            var attributes = fieldInfo.CustomAttributes;
            var attribute = attributes.FirstOrDefault(c => c.AttributeType == typeof (T));

            if (attribute != null && attribute.ConstructorArguments.Any())
            {
                return (string) attribute.ConstructorArguments[0].Value;
            }

            return Enum.GetName(value.GetType(), value);
        }

        public static Coordinates GetEnumCoordinates(object value)
        {
            var fieldInfo = value.GetType().
                                  GetField(value.ToString());
            var attributes = fieldInfo.CustomAttributes;
            var attribute = attributes.FirstOrDefault(c => c.AttributeType == typeof (CoordinatesAttribute));

            if (attribute != null && attribute.ConstructorArguments.Any())
            {
                return new Coordinates()
                       {
                           Latitude = (double) attribute.ConstructorArguments[0].Value,
                           Longitude = (double) attribute.ConstructorArguments[0].Value
                       };
            }

            return new Coordinates();
        }

        public static string GetEnumDefaultValue(object value)
        {
            return GetEnumAttribute<DefaultValueAttribute>(value);
        }

        public static string GetEnumDescription(object value)
        {
            return GetEnumAttribute<DescriptionAttribute>(value);
        }

        public static object GetJTokenObject(JToken item, string name)
        {
            return item.Value<object>(name) ?? null;
        }

        public static string GetJTokenString(JToken item, string name)
        {
            return item.Value<string>(name) ?? string.Empty;
        }

        /// <summary>
        ///     Checks for token validity
        /// </summary>
        /// <param name="jToken"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public static bool IsJTokenValid(JToken jToken)
        {
            var result = false;

            if (jToken != null)
            {
                if (!string.IsNullOrEmpty(jToken.ToString()))
                {
                    result = true;
                }
            }

            return result;
        }

        #endregion
    }
}

File: SourceCode/Projects/open.fda.gov/ApiDalc/Utilities.vb
﻿Imports ApiDalc.Enumerations

Public Module Utilities

#Region " Public Methods "

    ''' <summary>
    ''' Converts a string date to a Date Object
    ''' </summary>
    ''' <param name="dateString">21050102</param>
    ''' <param name="dateFormat">yyyyMMdd</param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function ConvertDateStringToDate(ByVal dateString As String, ByVal dateFormat As String) As Date

        If dateString Is Nothing OrElse dateString.Length = 0 Then
            Return Nothing
        End If

        Dim result As DateTime

        Select Case dateFormat
            Case "yyyyMMdd"
                result = DateTime.ParseExact(dateString, "yyyyMMdd", System.Globalization.CultureInfo.InvariantCulture)

            Case Else
                result = Nothing

        End Select

        Return result

    End Function


    ''' <summary>
    ''' Fixes String values
    ''' </summary>
    ''' <param name="str"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function AddBackslash(ByVal str As String) As String

        If str IsNot Nothing AndAlso Not str.EndsWith("\") Then
            str += "\"
        End If

        Return str

    End Function

    ''' <summary>
    ''' Fixes String values
    ''' </summary>
    ''' <param name="str"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function AddForwardSlash(ByVal str As String) As String

        If str IsNot Nothing AndAlso Not str.EndsWith("/") Then
            str += "/"
        End If

        Return str

    End Function

    ''' <summary>
    ''' Checks for token validity
    ''' </summary>
    ''' <param name="jToken"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function IsJTokenValid(ByVal jToken As Newtonsoft.Json.Linq.JToken) As Boolean

        Dim result As Boolean = False

        If jToken IsNot Nothing Then

            If Not String.IsNullOrEmpty(jToken.ToString) Then
                result = True
            End If

        End If

        Return result

    End Function

    ''' <summary>
    ''' Enumeration: Returns Description
    ''' </summary>
    ''' <param name="EnumConstant"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function GetEnumDescription(ByVal enumConstant As [Enum]) As String

        Dim result As String = String.Empty

        If enumConstant IsNot Nothing Then

            Dim fi As System.Reflection.FieldInfo = enumConstant.GetType().GetField(enumConstant.ToString())

            If fi IsNot Nothing Then

                Dim aattr() As System.ComponentModel.DescriptionAttribute = DirectCast(fi.GetCustomAttributes(GetType(System.ComponentModel.DescriptionAttribute), False), System.ComponentModel.DescriptionAttribute())

                If aattr.Length > 0 Then
                    result = aattr(0).Description
                Else
                    result = enumConstant.ToString()
                End If

            End If

        End If

        Return result

    End Function

    ''' <summary>
    ''' Enumeration: Returns Display Name
    ''' </summary>
    ''' <param name="EnumConstant"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function GetEnumDisplayName(ByVal enumConstant As [Enum]) As String

        Dim displayName As String = String.Empty

        If Not enumConstant Is Nothing Then

            Dim fi As System.Reflection.FieldInfo = enumConstant.GetType().GetField(enumConstant.ToString())
            Dim aattr() As System.ComponentModel.DisplayNameAttribute = DirectCast(fi.GetCustomAttributes(GetType(System.ComponentModel.DisplayNameAttribute), False), System.ComponentModel.DisplayNameAttribute())

            If aattr.Length > 0 Then
                displayName = aattr(0).DisplayName
            Else
                displayName = enumConstant.ToString()
            End If

        End If

        Return displayName

    End Function

    ''' <summary>
    ''' Enumeration: Returns Default Value
    ''' </summary>
    ''' <param name="EnumConstant"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function GetEnumDefaultValue(ByVal enumConstant As [Enum]) As String

        Dim defaultValue As String = String.Empty

        If enumConstant IsNot Nothing Then

            Dim fi As System.Reflection.FieldInfo = enumConstant.GetType().GetField(enumConstant.ToString())
            Dim aattr() As System.ComponentModel.DefaultValueAttribute = DirectCast(fi.GetCustomAttributes(GetType(System.ComponentModel.DefaultValueAttribute), False), System.ComponentModel.DefaultValueAttribute())

            If aattr.Length > 0 Then
                defaultValue = aattr(0).Value.ToString
            Else
                defaultValue = enumConstant.ToString()
            End If

        End If

        Return defaultValue

    End Function

#End Region

End Module
